internal/metrics/benchmark_test.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package benchmark_test
+
+import (
+	"log"
+	"testing"
+
+	prometheus "github.com/prometheus/client_golang/prometheus"
+	promExporter "go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/sdk/metric"
+
+	"github.com/jaegertracing/jaeger/internal/metrics/otelmetrics"
+	prom "github.com/jaegertracing/jaeger/internal/metrics/prometheus"
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+)
+
+func setupPrometheusFactory() metrics.Factory {
+	reg := prometheus.NewRegistry()
+	return prom.New(prom.WithRegisterer(reg))
+}
+
+func setupOTELFactory() metrics.Factory {
+	registry := prometheus.NewRegistry()
+	exporter, err := promExporter.New(promExporter.WithRegisterer(registry))
+	if err != nil {
+		log.Fatalf("Failed to create Prometheus exporter: %v", err)
+	}
+	meterProvider := metric.NewMeterProvider(
+		metric.WithReader(exporter),
+	)
+	return otelmetrics.NewFactory(meterProvider)
+}
+
+func benchmarkMetric(b *testing.B, factory metrics.Factory, metricType string) {
+	var metricObj interface{}
+	switch metricType {
+	case "counter":
+		metricObj = factory.Counter(metrics.Options{
+			Name: "test_counter",
+			Tags: map[string]string{"tag1": "value1"},
+		})
+	case "gauge":
+		metricObj = factory.Gauge(metrics.Options{
+			Name: "test_gauge",
+			Tags: map[string]string{"tag1": "value1"},
+		})
+	case "timer":
+		metricObj = factory.Timer(metrics.TimerOptions{
+			Name: "test_timer",
+			Tags: map[string]string{"tag1": "value1"},
+		})
+	case "histogram":
+		metricObj = factory.Histogram(metrics.HistogramOptions{
+			Name: "test_histogram",
+			Tags: map[string]string{"tag1": "value1"},
+		})
+	}
+
+	for i := 0; i < b.N; i++ {
+		switch m := metricObj.(type) {
```

Comment:
this introduces unnecessary overhead. There is little in common across the 4 types, instead of switch statements just define 4 different functions.